### PR TITLE
fix(consensus): add authority index bounds checking to prevent DoS panics

### LINKED ISSUE #24475: fix(consensus): Missing authority index bounds checking causes DoS vulnerability
## Summary

A Byzantine validator can crash honest validators by sending blocks with invalid authority indices, causing index-out-of-bounds panics.

## Problem Description

When processing blocks from other validators, the consensus layer accesses arrays using authority indices without bounds checking. A malicious validator can craft a block with an invalid author index (e.g., 255 when committee size is 4), causing honest validators to panic when they try to access arrays at these invalid indices.

### Affected Code Paths

- `DagState::contains_blocks()` - panics when checking if block with invalid author exists
- `DagState::get_cached_blocks_in_range()` - panics on `recent_refs_by_authority[authority]` access
- `DagState::get_last_cached_block_in_range()` - panics on array access
- `DagState::contains_cached_block_at_slot()` - panics on `evicted_rounds[slot.authority]` access
- `BlockManager::update_block_received_metrics()` - panics on `received_block_rounds[block.author()]` access

### Attack Scenario

1. Malicious validator crafts a block with `author = AuthorityIndex(255)`
2. Honest validator receives this block and tries to process it
3. Code accesses `recent_refs_by_authority[255]` where array size is only 4
4. Index out of bounds panic crashes the honest validator
5. Single Byzantine validator can crash all honest validators (DoS)

## Proposed Solution

Add defensive bounds checking before all array accesses using authority index:
- Add `is_valid_authority()` helper to DagState for bounds validation
- Add `get_hostname()` helper to BlockManager with fallback for invalid indices
- Return empty/false/None for invalid indices instead of panicking
- Log warnings for defense in depth monitoring

## Security Impact

- **Severity**: High (DoS vulnerability)
- **Attack Vector**: Network (Byzantine validator)
- **Impact**: All honest validators can be crashed by a single malicious validator

### PR DESCRIPTION
**Pull request by SherlockShemol**
## Description

This PR adds defensive bounds checking for authority indices in DagState and BlockManager to prevent index-out-of-bounds panics that could be triggered by malformed data from Byzantine validators.

Fixes #24475

## Problem

A Byzantine validator can craft a block with an invalid authority index (e.g., author index = 255 when committee size is only 4). When honest validators process such a block, they access arrays using this index without bounds checking, causing index-out-of-bounds panic and node crash.

This is a DoS (Denial of Service) vulnerability where a single malicious validator can crash all honest validators in the network.

## Affected Code Paths

- `DagState::contains_blocks()` - panics when checking if block exists
- `DagState::get_cached_blocks_in_range()` - panics on array access
- `DagState::contains_cached_block_at_slot()` - panics on array access
- `BlockManager::update_block_received_metrics()` - panics on array access

## Solution

Add defensive bounds checking before all array accesses using authority index:

- Add `is_valid_authority()` helper to DagState for bounds validation
- Add `get_hostname()` helper to BlockManager with fallback for invalid indices
- Add bounds checks to:
  - `get_last_block_for_authority()`
  - `get_cached_blocks_in_range()`
  - `get_last_cached_block_in_range()`
  - `contains_cached_block_at_slot()`
  - `contains_blocks()`
  - `add_commit()`
  - `update_block_metadata()`
  - `update_block_received_metrics()`
- Return empty/false/None for invalid indices instead of panicking
- Log warnings for defense in depth monitoring

## Test Plan

Added comprehensive test suite in `phantom_voter_test.rs`:
- `test_phantom_author_index` - Tests rejection of blocks with invalid author index
- `test_phantom_ancestor_index` - Tests rejection of blocks with invalid ancestor indices
- `test_index_validation_order` - Tests boundary conditions for index validation
- `test_block_manager_phantom_ancestor` - Tests BlockManager handling of phantom ancestors
- `test_random_phantom_indices` - Fuzz-like test with random large indices

All tests verify that the system does NOT panic on invalid indices.

## Release Notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates.

- [ ] Protocol: 
- [ ] Nodes (Validators and Full Nodes): 
- [x] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: consensus/core/src/tests/phantom_voter_test.rs [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+            // committee.authority(authority_index); // This would panic!
+        }
+    }
+
+    println!("\n=== Test Complete ===");
+}
+
+/// Test: BlockManager handles phantom indices in ancestors
+///
+/// VULNERABILITY DISCOVERED: DagState::contains_blocks() does not validate
```
Comment (mwtian) [inline consensus/core/src/tests/phantom_voter_test.rs:267]: Callers of any DagState API, such as DagState::contains_blocks(), are responsible for using BlockVerifier to validate the AuthorityIndex.

**Comment by mwtian** [comment]
Thanks for the work put into this. I think the tests on BlockVerifier can be useful. We can continue the discussion on https://github.com/MystenLabs/sui/issues/24475
